ProcFilter.cpp

```diff
@@ -0,0 +1,355 @@
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+#include <fstream>
+#include <system_error>
+#include <sys/types.h>
+#include <sys/stat.h> /* for stat() */
+#include <dirent.h>
+#include <pwd.h>
+#include <grp.h>
+#include <algorithm>
+#include <cctype>
+
+// This include file can only be included in ONE translation unit
+#include <auparse.h>
+
+#define MAX_ITERATION_DEPTH 10
+
+extern "C" {
+#include <dlfcn.h>
+}
+
+/*****************************************************************************
+ * Dynamicly load needed libaudit symbols
+ *
+ * There are two version of libaudit (libaudit0, and libaudit1) this makes it
+ * impossible to build once then run on all supported distro versions.
+ *
+ * But, since libauparse is available on all supported distros, and it also
+ * links to libaudit, all we need to do is call dlsym to get the function
+ * pointer(s) we need.
+ *
+ *****************************************************************************/
+
+
+
+/*****************************************************************************
+ ** ProcessInfo
+ *****************************************************************************/
+
+ProcessInfo::ProcessInfo(std::string description, int processId, int parentProcessId)
+{
+    name = description;
+    pid = processId;
+    ppid = parentProcessId;
+}
+
+const ProcessInfo ProcessInfo::Empty("0",0,0);
+
+inline bool ProcessInfo::operator==(const ProcessInfo& x) const
+{
+    return (pid == x.pid) && (ppid == x.ppid) && (name == x.name);
+}
+
+inline bool ProcessInfo::operator!=(const ProcessInfo& x) const
+{
+    return !(*this==x);
+}
+
+
+/*****************************************************************************
+ ** ProcFilter
+ *****************************************************************************/
+
+ProcFilter* ProcFilter::_instance = NULL;
+
+std::set<std::string> ProcFilter::_blocked_process_names;
+std::set<std::string> ProcFilter::_blocked_user_names;
+
+// -------- helper functions -----------------------------
+int ProcFilter::is_dir(std::string path)
+{
+    struct stat stat_buf;
+    stat(path.c_str(), &stat_buf);
+    int is_directory = S_ISDIR(stat_buf.st_mode);
+    return (is_directory ? 1: 0);
+}
+
+bool ProcFilter::is_number(const std::string& s)
+{
+    return !s.empty() &&
+           std::find_if(s.begin(), s.end(), [](char c) { return !std::isdigit(c); }) == s.end();
+}
+
+bool ProcFilter::is_process_running(int pid)
+{
+    std::stringstream ss_file;
+    ss_file << "/proc/" << pid << "/stat";
+    std::string file = ss_file.str();
+    struct stat stat_buf;
+    return (stat(file.c_str(), &stat_buf) == 0);
+}
+
+std::string ProcFilter::get_user_of_process(int pid)
+{
+    std::stringstream ss_file;
+    ss_file << "/proc/" << pid;
+    std::string file_name = ss_file.str();
+
+    struct stat stat_buf;
+    if(stat(file_name.c_str(), &stat_buf) == -1)
+    {
+        return std::string();
+    }
+    struct passwd *pw = getpwuid(stat_buf.st_uid);
```

Comment:
The UserDB already has this info, use it instead. See AuditEventProcessor.cpp:291 for an example of how to use it.

Reply:
suggested fix pushed